framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -70,6 +72,98 @@ public ValueTransfer(CFAbstractAnalysis<CFValue, CFStore, CFTransfer> analysis)
     }
 
     /**
+     * Returns a range of possible lengths for {@code subNode}, as casted to a String.
+     *
+     * @param stringLengths a list of string lengths for subNode, or null
+     */
+    private Range getStringLengthRange(
+            Node subNode, TransferInput<CFValue, CFStore> p, List<Integer> stringLengths) {
+        if (stringLengths != null) {
+            return ValueCheckerUtils.getRangeFromValues(stringLengths);
+        }
+
+        CFValue value = p.getValueOfSubNode(subNode);
+
+        AnnotationMirror arrayLenRangeAnno =
+                AnnotationUtils.getAnnotationByClass(value.getAnnotations(), ArrayLenRange.class);
+
+        if (arrayLenRangeAnno != null)
+            return ValueAnnotatedTypeFactory.getArrayLenRange(arrayLenRangeAnno);
+
+        // @StringVal, @UnknownVal, @BottomVal
+        AnnotationMirror topAnno =
+                AnnotationUtils.getAnnotationByClass(value.getAnnotations(), UnknownVal.class);
+        if (topAnno != null) {
+            return new Range(0, Integer.MAX_VALUE);
+        }
+        AnnotationMirror bottomAnno =
+                AnnotationUtils.getAnnotationByClass(value.getAnnotations(), BottomVal.class);
+        if (bottomAnno != null) {
+            return Range.NOTHING;
+        }
+
+        return new Range(0, Integer.MAX_VALUE);
+    }
+
+    /**
+     * Returns a list of possible lengths for {@code subNode}, as casted to a String. Returns null
+     * if {@code subNode}'s type is top/unknown. Returns an empty list if {@code subNode}'s type is
+     * bottom.
+     *
+     * @param stringValues a list of values for subNode, or null
+     */
+    private List<Integer> getStringLengths(
+            Node subNode, TransferInput<CFValue, CFStore> p, List<String> stringValues) {
+        if (stringValues != null) {
```

Comment:
For both this method and `getStringLengthRange` you have this pattern where something is passed in, and if it's non-null you make a very different decision. I think this would be more readable if you made this decision at the call sites, and remove the `stringValues` parameter.